content/blog/2022/2022-08-30-handling-timezones-in-spring.md

```diff
@@ -0,0 +1,798 @@
+---
+title: "Handling Timezones in a Spring Boot Application"
+categories: ["Spring"]
+date: 2022-08-30 00:00:00 +1100
+modified: 2022-08-30 00:00:00 +1100
+authors: ["ranjani"]
+description: "Handling timezones in a Spring Boot Application"
+image: images/stock/0111-clock-1200x628-branded.jpg
+url: spring-timezones
+---
+
+It is common to encounter applications that run on different timezones. Handling date and time operations consistently
+across multiple layers of an application can be tricky.
+
+In this article, we will try to understand the options available in Java and apply them in the context of a Spring
+application to effectively handle timezones.
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/spring-boot/timezones" %}}
+
+## Understanding GMT, UTC and DST
+
+- **Greenwich Mean Time (GMT)** is a timezone used in some parts of the world, mainly Europe and Africa.
+  It was recommended as the prime meridian of the world in 1884 and eventually became the basis for global system of
+  timezones.
+  However, the United Nations officially adopted UTC as a standard in 1972 since it was more accurate than GMT for
+  setting clocks.
+- **Universal Coordinated Time (UTC)** is not a timezone. It is a universally preferred standard that can be used to
+  display timezones.
+- **Daylight Savings Time (DST)** is the practice of setting clocks forward by one hour in the summer months
+  and back again in the fall, to make better use of natural daylight. **Neither GMT or UTC get affected by DST**. To
+  account for DST changes,
+  countries or states usually switch to another timezone. For instance in Australian summer, the states that observe DST
+  will move from
+  Australian Eastern Standard Time (AEST) to Australian Eastern Daylight Time (AEDT).
+
+Operations around dates, time and timezones can be confusing and prone to errors. To understand some problems around
+dates refer to this [article.](https://yourcalendricalfallacyis.com/)
+We will deep-dive into various aspects of handling timezones in the further sections.
+
+## Drawbacks of Legacy `java.util` Date Classes
+
+Let's look at a few reasons why you should avoid the date and time classes in the `java.util` package when developing
+applications.
+
+### Missing Timezone Information
+
+```java
+public class DeprecatedExamples {
+  @Test
+  public void testCurrentDate() {
+    Date now = new Date();
+    Date before = new Date(1661832030000L);
+    assertThat(now).isAfter(before);
+  }
+}
```

Comment:
Please go through the code examples one more time and add the class to the code like I did in this code snippet and the next few snippets. Please also check the code examples in the [preview](https://deploy-preview-502--reflectoring.netlify.app/spring-timezones/) and properly format it and introduce line breaks where the code would stick out of the right side and create a horizontal scroll bar) ... it helps to make indentation with 2 spaces.

Reply:
Sorry, I think my last edit broke the formatting of the code examples a bit, because I had IntelliJ format it for me. Once you have added the class to a code example, however, and then let IntelliJ format it for you (with tabs set to 2 spaces), then IntelliJ does all the work for you (IntelliJ can't format without the wrapping class, however ...).